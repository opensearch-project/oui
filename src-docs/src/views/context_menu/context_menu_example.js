/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Link } from 'react-router-dom';

import { renderToHtml } from '../../services';

import { GuideSectionTypes } from '../../components';

import {
  OuiCode,
  OuiContextMenu,
  OuiContextMenuItem,
  OuiContextMenuPanel,
} from '../../../../src/components';

import ContextMenu from './context_menu';
const contextMenuSource = require('./context_menu?raw');
const contextMenuHtml = renderToHtml(ContextMenu);
const contextMenuSnippet = `<OuiContextMenu
  initialPanelId={0}
  panels={[
    {
      id: 0,
      title: 'This is a context menu',
      items: [
        {
          name: 'Handle an onClick',
          icon: <OuiIcon type="search" size="m" />,
          onClick: () => {
            closePopover();
          },
        },
      ]
    }
  ]}
/>`;

import SinglePanel from './single_panel';
const singlePanelSource = require('./single_panel?raw');
const singlePanelHtml = renderToHtml(SinglePanel);
const singlePanelSnippet = `<OuiContextMenuPanel
  items={[
    <OuiContextMenuItem
      key=""
      onClick={}>
      This is a context menu item
    </OuiContextMenuItem>
  ]}
/>`;

import Small from './small';
const smallSizeSource = require('./small?raw');
const smallSizeHtml = renderToHtml(SinglePanel);
const smallSnippet = `<OuiContextMenuPanel
  size="s"
  items={items}
/>`;

import ContentPanel from './content_panel';
const contentPanelSource = require('./content_panel?raw');
const contentPanelHtml = renderToHtml(ContentPanel);
const contentPanelSnippet = `<OuiContextMenuPanel>
  <!-- React element as child -->
</OuiContextMenuPanel>
`;

import ContextMenuWithContent from './context_menu_with_content';
const contextMenuWithContentSource = require('./context_menu_with_content?raw');
const contextMenuWithContentHtml = renderToHtml(ContextMenuWithContent);

export const ContextMenuExample = {
  title: 'Context menu',
  sections: [
    {
      source: [
        {
          type: GuideSectionTypes.JS,
          code: contextMenuSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: contextMenuHtml,
        },
      ],
      text: (
        <p>
          <strong>OuiContextMenu</strong> is a nested menu system useful for
          navigating complicated trees. It lives within an{' '}
          <Link to="/layout/popover">
            <strong>OuiPopover</strong>
          </Link>{' '}
          which itself can be wrapped around any component (like a button in
          this example).
        </p>
      ),
      props: { OuiContextMenu, OuiContextMenuPanel, OuiContextMenuItem },
      snippet: contextMenuSnippet,
      demo: <ContextMenu />,
    },
    {
      title: 'Sizes',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: smallSizeSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: smallSizeHtml,
        },
      ],
      text: (
        <p>
          <strong>OuiContextMenu</strong> supports a small and medium{' '}
          <OuiCode>size</OuiCode>. The default size is medium,{' '}
          <OuiCode>m</OuiCode>, and should be used for most menus and major
          actions such as top application menus. Use the smaller size,{' '}
          <OuiCode>s</OuiCode>, for a more compressed version containing minor
          actions or repeated menus like in <strong>OuiTable</strong>{' '}
          pagination.
        </p>
      ),
      snippet: smallSnippet,
      demo: <Small />,
    },
    {
      title: 'With single panel',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: singlePanelSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: singlePanelHtml,
        },
      ],
      text: (
        <p>
          Use <strong>OuiContextMenuPanel</strong> for simple, non-nested
          context menus. The below pagination example has no nesting and no
          title.
        </p>
      ),
      snippet: singlePanelSnippet,
      demo: <SinglePanel />,
    },
    {
      title: 'Displaying custom elements',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: contentPanelSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: contentPanelHtml,
        },
      ],
      text: (
        <p>
          If you have custom content to show instead of a list of options, you
          can pass a React element as a child to{' '}
          <strong>OuiContextMenuPanel</strong>.
        </p>
      ),
      snippet: contentPanelSnippet,
      demo: <ContentPanel />,
    },
    {
      title: 'Using panels with mixed items & content',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: contextMenuWithContentSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: contextMenuWithContentHtml,
        },
      ],
      text: (
        <div>
          <p>
            Context menu panels can be passed React elements through the{' '}
            <OuiCode>content</OuiCode> prop instead of <OuiCode>items</OuiCode>.
            The panel will display your custom content without modification.
          </p>
          <p>
            If your panel contents have different widths or you need to ensure
            that a specific context menu panel has a certain width, add{' '}
            <OuiCode language="ts">width: [number of pixels]</OuiCode> to the
            panel tree.
          </p>
          <p>
            You can add separator lines in the <OuiCode>items</OuiCode> prop if
            you define an item as{' '}
            <OuiCode language="ts">{'{isSeparator: true}'}</OuiCode>. This will
            pass the rest of its fields as props to a{' '}
            <Link to="/layout/horizontal-rule">
              <strong>OuiHorizontalRule</strong>
            </Link>{' '}
            component.
          </p>
        </div>
      ),
      demo: <ContextMenuWithContent />,
    },
  ],
};
