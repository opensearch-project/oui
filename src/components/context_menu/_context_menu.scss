/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

$ouiContextMenuWidth: $ouiSize * 16;

.ouiContextMenu {
  width: $ouiContextMenuWidth;
  max-width: 100%;
  position: relative;
  overflow: hidden;
  transition: height $ouiAnimSpeedFast $ouiAnimSlightResistance;
  border-radius: $ouiBorderRadius;

  .ouiContextMenu__content {
    padding: $ouiSizeS;
  }
}

/**
  * 1. When there are multiple ContextMenuPanels, the ContextMenu will absolutely
  *    position them. ContextMenuPanel will break the layout of a Popover if it's
  *    absolutely positioned by default.
  */
.ouiContextMenu__panel {
  position: absolute; /* 1 */
}

.ouiContextMenu__icon {
  margin-right: $ouiSizeS;
}


/* OUI -> EUI Aliases */
$euiContextMenuWidth: $ouiContextMenuWidth;

/* End of Aliases */
