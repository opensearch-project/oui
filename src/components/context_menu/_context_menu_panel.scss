/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

.ouiContextMenuPanel {
  width: 100%;
  visibility: visible;
  outline-offset: -$ouiFocusRingSize;

  &:focus {
    outline: none; // Hide focus ring because of `tabindex=-1` on Safari
  }

  &.ouiContextMenuPanel-txInLeft {
    pointer-events: none;
    animation: ouiContextMenuPanelTxInLeft $ouiAnimSpeedNormal $ouiAnimSlightResistance;
  }

  &.ouiContextMenuPanel-txOutLeft {
    pointer-events: none;
    animation: ouiContextMenuPanelTxOutLeft $ouiAnimSpeedNormal $ouiAnimSlightResistance;
  }

  &.ouiContextMenuPanel-txInRight {
    pointer-events: none;
    animation: ouiContextMenuPanelTxInRight $ouiAnimSpeedNormal $ouiAnimSlightResistance;
  }

  &.ouiContextMenuPanel-txOutRight {
    pointer-events: none;
    animation: ouiContextMenuPanelTxOutRight $ouiAnimSpeedNormal $ouiAnimSlightResistance;
  }
}

.ouiContextMenuPanel--next {
  transform: translateX($ouiContextMenuWidth);
  visibility: hidden;
}

.ouiContextMenuPanel--previous {
  transform: translateX(-$ouiContextMenuWidth);
  visibility: hidden;
}

.ouiContextMenuPanelTitle {
  @include ouiPopoverTitle;
  padding: $ouiSizeM;
  width: 100%;
  text-align: left;
  outline-offset: -$ouiFocusRingSize;

  &:enabled:hover,
  &:enabled:focus {
    text-decoration: underline;
  }

  &--small {
    @include ouiPopoverTitle;
    padding: ($ouiSizeS * .75) $ouiSizeS;
  }
}

@keyframes ouiContextMenuPanelTxInLeft {
  0% {
    transform: translateX($ouiContextMenuWidth);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes ouiContextMenuPanelTxOutLeft {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateX(-$ouiContextMenuWidth);
  }
}

@keyframes ouiContextMenuPanelTxInRight {
  0% {
    transform: translateX(-$ouiContextMenuWidth);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes ouiContextMenuPanelTxOutRight {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateX($ouiContextMenuWidth);
  }
}
