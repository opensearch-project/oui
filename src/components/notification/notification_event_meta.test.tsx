/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { OuiNotificationEventMeta } from './notification_event_meta';
import { OuiContextMenuItem } from '../context_menu';

describe('OuiNotificationEventMeta', () => {
  test('is rendered', () => {
    const { container } = render(
      <OuiNotificationEventMeta
        id="id"
        type="Alert"
        time={<span>2 min ago</span>}
        eventName="eventName"
      />
    );

    expect(container).toMatchSnapshot();
  });

  describe('props', () => {
    test('severity is rendered', () => {
      const { container } = render(
        <OuiNotificationEventMeta
          id="id"
          type="Alert"
          time={<span>2 min ago</span>}
          severity="severity"
          eventName="eventName"
        />
      );

      expect(container).toMatchSnapshot();
    });

    test('badgeColor is rendered', () => {
      const { container } = render(
        <OuiNotificationEventMeta
          id="id"
          type="Alert"
          time={<span>2 min ago</span>}
          eventName="eventName"
          badgeColor="success"
        />
      );

      expect(container).toMatchSnapshot();
    });

    test('logoCloud is rendered', () => {
      const { container } = render(
        <OuiNotificationEventMeta
          id="id"
          type="Alert"
          time={<span>2 min ago</span>}
          iconType="logoCloud"
          eventName="eventName"
        />
      );

      expect(container).toMatchSnapshot();
    });

    test('contextMenuItems are rendered', () => {
      const contextMenuItems = [
        <OuiContextMenuItem key="contextMenuItemA">
          Mark as read
        </OuiContextMenuItem>,
        <OuiContextMenuItem key="contextMenuItemB">
          View messages like this
        </OuiContextMenuItem>,
        <OuiContextMenuItem key="contextMenuItemC">
          Don&apos;t notify me about this
        </OuiContextMenuItem>,
      ];

      render(
        <OuiNotificationEventMeta
          id="id"
          type="Alert"
          time={<span>2 min ago</span>}
          iconType="logoCloud"
          eventName="eventName"
          onOpenContextMenu={() => contextMenuItems}
        />
      );

      // Initially, context menu items should not be visible
      expect(screen.queryByText('Mark as read')).not.toBeInTheDocument();

      // Click the button to open the context menu
      act(() => {
        fireEvent.click(screen.getByTestId('id-notificationEventMetaButton'));
      });

      // Now the context menu items should be visible
      expect(screen.getByText('Mark as read')).toBeInTheDocument();
      expect(screen.getByText('View messages like this')).toBeInTheDocument();
      expect(
        screen.getByText("Don't notify me about this")
      ).toBeInTheDocument();

      // Take a snapshot of the entire container which should include the portal
      expect(document.body).toMatchSnapshot();
    });
  });
});
