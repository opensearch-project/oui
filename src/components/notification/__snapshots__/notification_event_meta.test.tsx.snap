// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiNotificationEventMeta is rendered 1`] = `
<div>
  <div
    class="ouiNotificationEventMeta"
  >
    <div
      class="ouiNotificationEventMeta__section"
    >
      <span
        class="ouiBadge ouiBadge--hollow ouiBadge--iconLeft ouiNotificationEventMeta__badge"
        title="Alert"
      >
        <span
          class="ouiBadge__content"
        >
          <span
            class="ouiBadge__text"
          >
            Alert
          </span>
        </span>
      </span>
    </div>
    <div
      class="ouiNotificationEventMeta__section"
    >
      <span
        class="ouiNotificationEventMeta__time"
      >
        <span>
          2 min ago
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiNotificationEventMeta props badgeColor is rendered 1`] = `
<div>
  <div
    class="ouiNotificationEventMeta"
  >
    <div
      class="ouiNotificationEventMeta__section"
    >
      <span
        class="ouiBadge ouiBadge--iconLeft ouiNotificationEventMeta__badge"
        style="background-color: rgb(109, 204, 177); color: rgb(0, 0, 0);"
        title="Alert"
      >
        <span
          class="ouiBadge__content"
        >
          <span
            class="ouiBadge__text"
          >
            Alert
          </span>
        </span>
      </span>
    </div>
    <div
      class="ouiNotificationEventMeta__section"
    >
      <span
        class="ouiNotificationEventMeta__time"
      >
        <span>
          2 min ago
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiNotificationEventMeta props contextMenuItems are rendered 1`] = `
<body>
  <div>
    <div
      class="ouiNotificationEventMeta ouiNotificationEventMeta--hasContextMenu"
    >
      <div
        class="ouiNotificationEventMeta__section"
      >
        <span
          aria-hidden="true"
          class="ouiNotificationEventMeta__icon"
          data-ouiicon-type="logoCloud"
        />
        <span
          class="ouiBadge ouiBadge--hollow ouiBadge--iconLeft ouiNotificationEventMeta__badge"
          title="Alert"
        >
          <span
            class="ouiBadge__content"
          >
            <span
              class="ouiBadge__text"
            >
              Alert
            </span>
          </span>
        </span>
      </div>
      <div
        class="ouiNotificationEventMeta__section"
      >
        <span
          class="ouiNotificationEventMeta__time"
        >
          <span>
            2 min ago
          </span>
        </span>
      </div>
      <div
        class="ouiNotificationEventMeta__contextMenuWrapper"
      >
        <div
          class="ouiPopover ouiPopover--anchorLeftUp"
          id="generated-id"
        >
          <div
            class="ouiPopover__anchor"
          >
            <button
              aria-controls="generated-id"
              aria-expanded="true"
              aria-haspopup="true"
              aria-label="Menu for eventName"
              class="ouiButtonIcon ouiButtonIcon--subdued ouiButtonIcon--empty ouiButtonIcon--xSmall"
              data-test-subj="id-notificationEventMetaButton"
              type="button"
            >
              <span
                aria-hidden="true"
                class="ouiButtonIcon__icon"
                color="inherit"
                data-ouiicon-type="boxesVertical"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="-1"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="-1"
    />
    <div
      data-focus-lock-disabled="disabled"
    >
      <div
        aria-describedby="generated-id"
        aria-live="off"
        aria-modal="true"
        class="ouiPanel ouiPanel--borderRadiusMedium ouiPanel--plain ouiPanel--noShadow ouiPopover__panel ouiPopover__panel--left"
        data-autofocus="true"
        role="dialog"
        style="top: -22px; left: -16px; will-change: transform, opacity; z-index: 2000;"
        tabindex="0"
      >
        <div
          class="ouiPopover__panelArrow ouiPopover__panelArrow--left"
          style="left: 0px; top: 10px;"
        />
        <p
          class="ouiScreenReaderOnly"
          id="generated-id"
        >
          You are in a dialog. To close this dialog, hit escape.
        </p>
        <div>
          <div>
            <div
              class="ouiContextMenuPanel"
              tabindex="-1"
            >
              <div>
                <div>
                  <button
                    class="ouiContextMenuItem"
                    type="button"
                  >
                    <span
                      class="ouiContextMenu__itemLayout"
                    >
                      <span
                        class="ouiContextMenuItem__text"
                      >
                        Mark as read
                      </span>
                    </span>
                  </button>
                  <button
                    class="ouiContextMenuItem"
                    type="button"
                  >
                    <span
                      class="ouiContextMenu__itemLayout"
                    >
                      <span
                        class="ouiContextMenuItem__text"
                      >
                        View messages like this
                      </span>
                    </span>
                  </button>
                  <button
                    class="ouiContextMenuItem"
                    type="button"
                  >
                    <span
                      class="ouiContextMenu__itemLayout"
                    >
                      <span
                        class="ouiContextMenuItem__text"
                      >
                        Don't notify me about this
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="-1"
    />
  </div>
</body>
`;

exports[`OuiNotificationEventMeta props logoCloud is rendered 1`] = `
<div>
  <div
    class="ouiNotificationEventMeta"
  >
    <div
      class="ouiNotificationEventMeta__section"
    >
      <span
        aria-hidden="true"
        class="ouiNotificationEventMeta__icon"
        data-ouiicon-type="logoCloud"
      />
      <span
        class="ouiBadge ouiBadge--hollow ouiBadge--iconLeft ouiNotificationEventMeta__badge"
        title="Alert"
      >
        <span
          class="ouiBadge__content"
        >
          <span
            class="ouiBadge__text"
          >
            Alert
          </span>
        </span>
      </span>
    </div>
    <div
      class="ouiNotificationEventMeta__section"
    >
      <span
        class="ouiNotificationEventMeta__time"
      >
        <span>
          2 min ago
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiNotificationEventMeta props severity is rendered 1`] = `
<div>
  <div
    class="ouiNotificationEventMeta"
  >
    <div
      class="ouiNotificationEventMeta__section"
    >
      <span
        class="ouiBadge ouiBadge--hollow ouiBadge--iconLeft ouiNotificationEventMeta__badge"
        title="Alert: severity"
      >
        <span
          class="ouiBadge__content"
        >
          <span
            class="ouiBadge__text"
          >
            Alert: severity
          </span>
        </span>
      </span>
    </div>
    <div
      class="ouiNotificationEventMeta__section"
    >
      <span
        class="ouiNotificationEventMeta__time"
      >
        <span>
          2 min ago
        </span>
      </span>
    </div>
  </div>
</div>
`;
